frontend/sw360-portlet/src/main/java/org/eclipse/sw360/portal/users/UserUtils.java

```diff
@@ -58,32 +58,49 @@ public UserUtils() {
+    public static <T> org.eclipse.sw360.datahandler.thrift.users.User synchronizeUserWithDatabase(
+            T source, ThriftClients thriftClients, Supplier<String> emailSupplier,
+            Supplier<String> extIdSupplier, BiConsumer<org.eclipse.sw360.datahandler.thrift.users.User, T> synchronizer) {
         UserService.Iface client = thriftClients.makeUserClient();
 
-        org.eclipse.sw360.datahandler.thrift.users.User thriftUser = null;
+        org.eclipse.sw360.datahandler.thrift.users.User existingThriftUser = null;
 
+        String email = emailSupplier.get();
         try {
-            thriftUser = client.getByEmail(emailSupplier.get());
+            existingThriftUser = client.getByEmailOrExternalId(email, extIdSupplier.get());
         } catch (TException e) {
             //This occurs for every new user, so there is not necessarily something wrong
-            log.trace("User does not exist in DB yet.");
+            log.trace("User not found by email or external ID");
         }
 
+        org.eclipse.sw360.datahandler.thrift.users.User resultUser = null;
         try {
-            if (thriftUser == null) {
+            if (existingThriftUser == null) {
                 log.info("Creating new user.");
-                thriftUser = new org.eclipse.sw360.datahandler.thrift.users.User();
-                synchronizer.accept(thriftUser, source);
-                client.addUser(thriftUser);
+                resultUser = new org.eclipse.sw360.datahandler.thrift.users.User();
+                synchronizer.accept(resultUser, source);
+                client.addUser(resultUser);
             } else {
-                synchronizer.accept(thriftUser, source);
-                client.updateUser(thriftUser);
+                resultUser = existingThriftUser;
+                if (!existingThriftUser.getEmail().equals(email)) { // email has changed
+                    resultUser.setFormerEmailAddresses(prepareFormerEmailAddresses(existingThriftUser, email));
+                }
+                synchronizer.accept(resultUser, source);
+                client.updateUser(resultUser);
             }
         } catch (TException e) {
             log.error("Thrift exception when saving the user", e);
         }
-        return thriftUser;
+        return resultUser;
+    }
+
+    @NotNull
+    private static Set<String> prepareFormerEmailAddresses(org.eclipse.sw360.datahandler.thrift.users.User thriftUser, String email) {
+        Set<String> formerEmailAddresses = nullToEmptySet(thriftUser.getFormerEmailAddresses()).stream()
+                .filter(e -> !e.equals(email)) // make sure the current email is not in the former addresses
+                .collect(Collectors.toCollection(HashSet::new));
```

Comment:
Because I need to ensure that the resulting `Set` is mutable, or else I won't be able to add to the `Set` in the next line. See [javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/stream/Collectors.html#toSet--)